File: streampark-flink/streampark-flink-connector/streampark-flink-connector-http/src/main/scala/org/apache/streampark/flink/connector/http/internal/HttpWriterTask.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.http.internal

import org.apache.streampark.common.util.{JsonUtils, Logger}
import org.apache.streampark.flink.connector.conf.ThresholdConf
import org.apache.streampark.flink.connector.failover.{FailoverWriter, SinkRequest}

import io.netty.handler.codec.http.HttpHeaders
import org.apache.http.client.methods._
import org.asynchttpclient.{AsyncHttpClient, ListenableFuture, Request, Response}

import java.util
import java.util.concurrent.{BlockingQueue, ExecutorService, TimeUnit}

import org.apache.streampark.common.util.Implicits._
import scala.util.Try

case class HttpWriterTask(
    id: Int,
    thresholdConf: ThresholdConf,
    asyncHttpClient: AsyncHttpClient,
    header: Map[String, String],
    queue: BlockingQueue[SinkRequest],
    callbackService: ExecutorService)
  extends Runnable
  with AutoCloseable
  with Logger {

  @volatile var isWorking = false

  val httpMethods: List[String] = List[String](
    HttpGet.METHOD_NAME,
    HttpPost.METHOD_NAME,
    HttpPut.METHOD_NAME,
    HttpPatch.METHOD_NAME,
    HttpDelete.METHOD_NAME,
    HttpOptions.METHOD_NAME,
    HttpTrace.METHOD_NAME
  )

  val failoverWriter: FailoverWriter =
    new FailoverWriter(thresholdConf.storageType, thresholdConf.getFailoverConfig)

  def buildRequest(url: String): Request = {

    val method = httpMethods.filter(x => url.startsWith(x)).head

    val uriAndParams = url.drop(method.length + 3).split("\\?")

    val uri = uriAndParams.head

    val builder = method match {
      case HttpGet.METHOD_NAME => asyncHttpClient.prepareGet(uri)
      case HttpDelete.METHOD_NAME => asyncHttpClient.prepareDelete(uri)
      case HttpOptions.METHOD_NAME => asyncHttpClient.prepareOptions(uri)
      case HttpTrace.METHOD_NAME => asyncHttpClient.prepareTrace(uri)
      case HttpPost.METHOD_NAME => asyncHttpClient.preparePost(uri)
      case HttpPatch.METHOD_NAME => asyncHttpClient.preparePatch(uri)
      case HttpPut.METHOD_NAME => asyncHttpClient.preparePut(uri)
    }
    if (header != null && header.nonEmpty) {
      header.foreach { case (k, v) => builder.setHeader(k, v) }
    }

    Try(uriAndParams(1).trim).getOrElse(null) match {
      case null =>
      case params =>
        val paramMap = new util.HashMap[String, String]()
        params
          .split("&")
          .foreach(
            x => {
              val param = x.split("=")
              paramMap.put(param.head, param.last)
            })
        if (paramMap.nonEmpty) {
          builder.setHeader(HttpHeaders.Names.CONTENT_TYPE, HttpHeaders.Values.APPLICATION_JSON)
          val json = JsonUtils.write(paramMap)
          builder.setBody(json.getBytes)
        }
    }
    builder.setRequestTimeout(thresholdConf.timeout).build()
  }

  override def run(): Unit =
    try {
      isWorking = true
      logInfo(s"Start writer task, id = $id")
      while (isWorking || queue.nonEmpty) {
        val req = queue.poll(100, TimeUnit.MILLISECONDS)
        if (req != null) {
          val url = req.records.head
          val sinkRequest = SinkRequest(List(url), req.attemptCounter)
          val request = buildRequest(url)
          val whenResponse = asyncHttpClient.executeRequest(request)
          val callback = respCallback(whenResponse, sinkRequest)
          whenResponse.addListener(callback, callbackService)
          if (req.attemptCounter > 0) {
            logInfo(s"get retry url from queue,attemptCounter:${req.attemptCounter}")
          }
        }
      }
    } catch {
      case e: Exception =>
        logError("Error while inserting data", e)
        throw new RuntimeException(e)
    } finally {
      logInfo(s"Task id = $id is finished")
    }

  def respCallback(whenResponse: ListenableFuture[Response], sinkRequest: SinkRequest): Runnable =
    new Runnable {
      override def run(): Unit = {
        Try(whenResponse.get()).getOrElse(null) match {
          case null =>
            logError(
              s"""Error HttpSink executing callback, params = $thresholdConf,can not get Response. """)
            handleFailedResponse(null, sinkRequest)
          case resp if resp.getStatusCode != 200 =>
            logError(
              s"""Error HttpSink executing callback, params = $thresholdConf, StatusCode = ${resp.getStatusCode} """)
            handleFailedResponse(resp, sinkRequest)
          case _ =>
        }
      }
    }

  /**
   * if send data to Http Failed, retry maxRetries, if still failed,flush data to failoverStorage
   *
   * @param response
   * @param sinkRequest
   */
  def handleFailedResponse(response: Response, sinkRequest: SinkRequest): Unit =
    try {
      if (sinkRequest.attemptCounter >= thresholdConf.maxRetries) {
        failoverWriter.write(
          sinkRequest.copy(records = sinkRequest.records.map(_.replaceFirst("^[A-Z]+///", ""))))
        logWarn(
          s"""Failed to send data to Http, Http response = $response. Ready to flush data to ${thresholdConf.storageType}""")
      } else {
        sinkRequest.incrementCounter()
        logWarn(
          s"Next attempt to send data to Http, table = ${sinkRequest.table}, buffer size = ${sinkRequest.size}, current attempt num = ${sinkRequest.attemptCounter}, max attempt num = ${thresholdConf.maxRetries}, response = $response")
        queue.put(sinkRequest)
      }
    } catch {
      case e: Exception => new RuntimeException(s"[StreamPark] handleFailedResponse,error:$e")
    }

  override def close(): Unit = {
    isWorking = false
    failoverWriter.close()
  }
}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-http/src/main/java/org/apache/streampark/flink/connector/http/internal/HttpWriterTask.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.http.internal;
import org.apache.streampark.common.util.JsonUtils;
import org.apache.streampark.flink.connector.conf.ThresholdConf;
import org.apache.streampark.flink.connector.failover.FailoverWriter;
import org.apache.streampark.flink.connector.failover.SinkRequest;
import org.apache.http.client.methods.*;
import org.asynchttpclient.*;
import io.netty.handler.codec.http.HttpHeaders;
import org.slf4j.Logger; import org.slf4j.LoggerFactory;
import java.util.*; import java.util.concurrent.*;
public class HttpWriterTask implements Runnable, AutoCloseable {
    private static final Logger LOG = LoggerFactory.getLogger(HttpWriterTask.class);
    private static final List<String> HTTP_METHODS = Arrays.asList(
        HttpGet.METHOD_NAME, HttpPost.METHOD_NAME, HttpPut.METHOD_NAME, HttpPatch.METHOD_NAME,
        HttpDelete.METHOD_NAME, HttpOptions.METHOD_NAME, HttpTrace.METHOD_NAME);
    private final int id; private final ThresholdConf thresholdConf; private final AsyncHttpClient asyncHttpClient;
    private final Map<String,String> header; private final BlockingQueue<SinkRequest> queue;
    private final ExecutorService callbackService; private volatile boolean isWorking = false;
    private final FailoverWriter failoverWriter;
    public HttpWriterTask(int id, ThresholdConf tc, AsyncHttpClient client, Map<String,String> header, BlockingQueue<SinkRequest> queue, ExecutorService cb) {
        this.id = id; this.thresholdConf = tc; this.asyncHttpClient = client; this.header = header; this.queue = queue; this.callbackService = cb;
        failoverWriter = new FailoverWriter(thresholdConf.storageType, thresholdConf.getFailoverConfig());
    }
    @Override public void run() {
        try { isWorking = true;
            while (isWorking || !queue.isEmpty()) {
                SinkRequest req = queue.poll(100, TimeUnit.MILLISECONDS);
                if (req != null && !req.getRecords().isEmpty()) {
                    String url = req.getRecords().get(0);
                    SinkRequest sinkRequest = new SinkRequest(Collections.singletonList(url), req.getAttemptCounter());
                    ListenableFuture<Response> future = asyncHttpClient.executeRequest(buildRequest(url));
                    future.addListener(() -> { try { Response resp = future.get(); if (resp == null || resp.getStatusCode() != 200) handleFailedResponse(resp, sinkRequest); } catch (Exception e) { handleFailedResponse(null, sinkRequest); }}, callbackService);
                }
            }
        } catch (Exception e) { throw new RuntimeException(e); }
        finally { LOG.info("Task id = {} is finished", id); }
    }
    Request buildRequest(String url) {
        String method = HTTP_METHODS.stream().filter(url::startsWith).findFirst().orElse(HttpGet.METHOD_NAME);
        String[] uriAndParams = url.substring(method.length() + 3).split("\\?", 2);
        String uri = uriAndParams[0];
        BoundRequestBuilder builder;
        switch (method) {
            case HttpDelete.METHOD_NAME: builder = asyncHttpClient.prepareDelete(uri); break;
            case HttpOptions.METHOD_NAME: builder = asyncHttpClient.prepareOptions(uri); break;
            case HttpTrace.METHOD_NAME: builder = asyncHttpClient.prepareTrace(uri); break;
            case HttpPost.METHOD_NAME: builder = asyncHttpClient.preparePost(uri); break;
            case HttpPatch.METHOD_NAME: builder = asyncHttpClient.preparePatch(uri); break;
            case HttpPut.METHOD_NAME: builder = asyncHttpClient.preparePut(uri); break;
            default: builder = asyncHttpClient.prepareGet(uri);
        }
        if (header != null) header.forEach(builder::setHeader);
        if (uriAndParams.length > 1) {
            Map<String,String> paramMap = new HashMap<>();
            for (String x : uriAndParams[1].trim().split("&")) {
                String[] param = x.split("=", 2);
                if (param.length == 2) paramMap.put(param[0], param[1]);
            }
            if (!paramMap.isEmpty()) {
                builder.setHeader(HttpHeaders.Names.CONTENT_TYPE, HttpHeaders.Values.APPLICATION_JSON);
                try {
                    builder.setBody(JsonUtils.write(paramMap).getBytes());
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }
        }
        return builder.setRequestTimeout(thresholdConf.timeout).build();
    }
    void handleFailedResponse(Response response, SinkRequest sinkRequest) {
        try {
            if (sinkRequest.getAttemptCounter() >= thresholdConf.maxRetries) {
                List<String> cleaned = new ArrayList<>();
                for (String r : sinkRequest.getRecords()) cleaned.add(r.replaceFirst("^[A-Z]+///", ""));
                failoverWriter.write(new SinkRequest(cleaned, sinkRequest.getAttemptCounter()));
            } else { sinkRequest.incrementCounter(); queue.put(sinkRequest); }
        } catch (Exception e) { throw new RuntimeException(e); }
    }
    @Override public void close() { isWorking = false; failoverWriter.close(); }
}
